### ISSUE TITLE
False positive error message caused by MRO parsing 

### ISSUE THREAD

**Issue by AndrewUshakov** [python/mypy#5882]
For the working code below (Python 3.7.1):

```Python
class A:

    def __init__(self, i: int) -> None:
        print('A.__init__()')


class B:
    ...


class C(B):

    def __init__(self) -> None:
        ...


class D(C, A):

    def __init__(self) -> None:
        super().__init__()
        super(C, self).__init__(1)   # tst85a.py:21: error: Too many arguments for "__init__" of "C"  


if __name__ == '__main__':
    help(D)
    d = D()
```
which output is:
```
class D(C, A)
 |  D() -> None
 |  
 |  Method resolution order:
 |      D
 |      C
 |      B
 |      A
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __init__(self) -> None
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from B:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)

A.__init__()

Process finished with exit code 0
```
mypy, just installed from github:
```
(.venv) D:\My Documents\PycharmProjects\tst>py -m mypy  -V
mypy 0.650+dev.e5e453233f347c3f18b3ae4e98e5276c8ce622ab
```
generates false positive error:
```
(.venv) D:\My Documents\PycharmProjects\tst>py -m mypy --strict tst85a.py
tst85a.py:21: error: Too many arguments for "__init__" of "C"
```
for the last line of ```D.__init__()``` while MRO is D,C,B,A,object,  class B does not have ```__init__```, so ```A.__init__()``` was really called as expected (see output above).

**Comment by gvanrossum** [comment]
If you're sure that `A.__init__` will be called, you should just call it without using `super`. The MRO is somewhat dynamic, in that I could create a class that inserted a different class in the MRO between B and A:
```py
class A1(A): ...
class B1(B, A1): ...
class D1(D, B1): ...
print(D1.__mro__)  # D1, D, C, B1, B, A1, A, object
```
And if A1 were to define an `__init__` method, `super(C, self).__init__(1)` would call `A1.__init__`, not the one you expected.

This is of course a roundabout way of saying that `super().__init__` calls are *always* unsafe, because `__init__` does not need to follow Liskov. If you want something a little more type-safe, use a different method name.

I do see a problem though: the error message `for "__init__" of "C"` is misleading, since the method being called is expressly *not* a method of C -- it is a method of the first class after C in the MRO of `self.__class__`.

**Comment by AndrewUshakov** [comment]
Call of the parent directly isn't a universal solution, just because this approach probably also uses MRO (and parent name is just a start point to search): For the code below no run time (and Pycharm/mypy) error will be generated for missing ```C.__init__```, but ```A.__init__``` will be called.
```Python
class A:

    def __int__(self) -> None:
        print('A.__init__()...')


class B:
    def __init__(self) -> None:
        print('A.__init__()...')


class C(A, B):
    ...


class D(C, B):

    def __init__(self) -> None:
        C.__init__(self)


if __name__ == '__main__':
    help(C)
    help(D)
    d = D()
```

if replace ```class C(A, B)``` by ```class C(A)``, nothing will be called, and I do not understand why.

**Comment by gvanrossum** [comment]
That's because if C doesn't override x, but a base class of C defines x, `C.x` will reference the x on the base class. IOW in this case `C.__init__` is `A.__init__`. That's how inheritance works.

**Comment by AndrewUshakov** [comment]
Ops, after reading your last comment I found 2 obvious errors in my code from the previous message. First: in the class A I defined not ```__init__``` by ```__int__``` and in the class B printed confusing message :) Thank you!

**Comment by gvanrossum** [comment]
Let's close this then.
